src/java/org/apache/cassandra/streaming/StreamManager.java

```diff
@@ -252,18 +255,36 @@ public void onRegister(StreamResultFuture result)
         }
     };
 
+    protected void addStreamingStateAgain(StreamingState state)
+    {
+        if (!DatabaseDescriptor.getStreamingStatsEnabled())
+            return;
+        states.put(state.id(), state);
+    }
+
     public StreamManager()
     {
         DurationSpec.LongNanosecondsBound duration = DatabaseDescriptor.getStreamingStateExpires();
         long sizeBytes = DatabaseDescriptor.getStreamingStateSize().toBytes();
-        long numElements = sizeBytes / StreamingState.ELEMENT_SIZE;
-        logger.info("Storing streaming state for {} or for {} elements", duration, numElements);
+        logger.info("Storing streaming state for {} or for weight {}", duration, sizeBytes);
         states = CacheBuilder.newBuilder()
                              .expireAfterWrite(duration.quantity(), duration.unit())
-                             .maximumSize(numElements)
+                             .maximumWeight(sizeBytes)
+                             .weigher(new StreamingStateWeigher())
                              .build();
     }
 
+    private static class StreamingStateWeigher implements Weigher<TimeUUID,StreamingState>
+    {
+        @Override
+        public int weigh(TimeUUID key, StreamingState val)
+        {
+            long costOfStreamingState = val.unsharedHeapSize() + TimeUUID.sizeInBytes();
```

Comment:
`TimeUUID.sizeInBytes()` this is the serialized size and not the object overhead.  Please compute this like we do `EMPTY` for other types